Integrate the CMake Debugger

gcampbell-msft opened the pull request:
<!-- Thanks for your contribution! To make things easier, please fill out the template below. -->

<!-- Please delete any unused sections. -->

<!-- Delete the following heading if there is no corresponding issue -->
## This change addresses item #3093 

NOTE: This PR shouldn't merge or be officially released until there is release candidate available that supports the CMake Debugger.

### This integrates the CMake debugger

The following changes are proposed:

- Adds various entry points to configure with the CMake Debugger
- This includes entry points in the command pallete and many entry points in the Project Outline. Overall, anywhere that you can configure, you now can also configure with CMake Debugger.
- Variables, call stack, breakpoints, cache variables, and more are supported!

## The purpose of this change
Integrate the CMake Debugger. 

<!-- If not linked to an issue, describe the purpose of this change. Otherwise,
delete this section. -->

## Other Notes/Information
This only allows for "Configure with CMake Debugger" entrypoints when the cmake exe that the extension is using has the debugger capabilities enabled. This implementation uses a vscode extension context variable to enable or disable the entrypoints. It uses the cmake.exe -E capabilities JSON output to determine if the cmake.exe supports the debugger.

<!-- Write whatever you feel is relevant -->

gcampbell-msft commented:
Probably needs a changelog update before merging.

gcampbell-msft commented:
Investigate killing the configure process when the debug session is forcibly stopped.

benmcmorran commented:
@gcampbell-msft This should be ready for testing with the 3.27 release candidates, [available here](https://cmake.org/download/).

gcampbell-msft commented:
> @gcampbell-msft This should be ready for testing with the 3.27 release candidates, [available here](https://cmake.org/download/).

I tested with the release candidate and things seem to work well! I think it's ready to go into pre-release, pending further PR review.

gcampbell-msft commented:
@benmcmorran @snehara99 Ping for review. 👍

bobbrow commented on CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 
+## 1.15
+Features:
+- Added support for the CMake Debugger! [#3093](https://github.com/microsoft/vscode-cmake-tools/issues/3093)
```
I can understand the excitement, but we don't usually put exclamations points in the changelog.  It seems more like something for the blogpost.

gcampbell-msft replied:
Haha fair point, I'll adjust

bobbrow commented on src/drivers/cmakeFileApiDriver.ts:
```diff
@@ -245,11 +253,21 @@ export class CMakeFileApiDriver extends CMakeDriver {
             });
             this.configureProcess = child;
+
+            // TODO: Before PR'ing and merging, I should add implementation to invoke cmake under the hood with the "-E capabilities" flag
```
Will this be addressed?

gcampbell-msft replied:
Ah, good catch, I've already implemented this, I'll remove.

bobbrow replied:
It looks like you did something in cmakeExecutable.ts already. Maybe make it available here?

benmcmorran commented on src/debug/debuggerConfigureDriver.ts:
```diff
@@ -0,0 +1,26 @@
+import { CMakeOutputConsumer } from "@cmt/diagnostics/cmake";
+import * as vscode from "vscode";
+
+export const debuggerPipeName = "\\\\.\\\\pipe\\\\cmake-debugger-pipe";
```
Using a constant name for the pipe makes it impossible to have multiple CMake debug sessions running at the same time.

gcampbell-msft replied:
Fixing.

benmcmorran commented on package.json:
```diff
@@ -223,6 +229,7 @@
         }
       },
       {
+
```
Nit: unneeded blank line

gcampbell-msft replied:
Fixing

benmcmorran commented on package.json:
```diff
@@ -665,6 +712,17 @@
+    "debuggers": [
+        {
+            "type": "cmake",
+            "label": "CMake Debugger"
```
Can this be localized?

gcampbell-msft replied:
Changing.

benmcmorran commented on src/debug/debuggerConfigureDriver.ts:
```diff
@@ -0,0 +1,26 @@
+export const debuggerPipeName = "\\\\.\\\\pipe\\\\cmake-debugger-pipe";
+
+export async function startConfigureDebugger(outputConsumer: CMakeOutputConsumer): Promise<void> {
+    if (outputConsumer instanceof CMakeOutputConsumer) {
```
Is this check necessary? If so, then the type annotation on the parameter should be relaxed to capture the range of types we actually expect to be passed to this function.

gcampbell-msft replied:
Good point. I only expect it to ever be a CMakeOutputConsumer, so I'll remove the check.

gcampbell-msft replied:
Removing this check.

benmcmorran commented:
One other high-level thought -- since users can now create `launch.json` configurations to debug CMake, I can imagine use cases for debugging in [script mode](https://cmake.org/cmake/help/latest/manual/cmake.1.html#run-a-script) instead of an entire project. Would it be possible to add debug configuration settings to enable this scenario?

gcampbell-msft commented on package.json:
```diff
@@ -232,6 +239,12 @@
+        "command": "cmake.outline.configureWithDebugger",
+        "title": "%cmake-tools.command.cmake.configureWithDebugger.title%",
+        "icon": "$(debug)",
+        "enablement": "vscode-cmake-tools.cmakeDebuggerAvailable == 'true'"
```
Changing.

gcampbell-msft commented on src/cmake/cmakeExecutable.ts:
```diff
@@ -44,6 +52,17 @@ export async function getCMakeExecutableInformation(path: string): Promise<CMake
+                console.assert(debuggerPresent.stdout);
+                const stdoutJson = JSON.parse(debuggerPresent.stdout);
+                cmake.isDebuggerSupported = stdoutJson["debugger"];
+                await vscode.commands.executeCommand(
```
Changing.

gcampbell-msft commented on src/cmakeProject.ts:
```diff
@@ -1334,6 +1357,21 @@ export class CMakeProject {
+                                } else if (result !== 0 && (await this.getCMakeExecutable()).isDebuggerSupported) {
+                                    const yesButtonTitle: string = localize(
+                                        "yes.configureWithDebugger.button",
+                                        "Yes"
```
Changing.

gcampbell-msft commented on src/debug/debuggerConfigureDriver.ts:
```diff
@@ -0,0 +1,26 @@
+                    return resolve();
+                }
+                setTimeout(waitForDebuggerClientMessage, 50);
+            })();
```
Changing.

gcampbell-msft commented on src/debug/debuggerConfigureDriver.ts:
```diff
@@ -0,0 +1,26 @@
+        await new Promise<void>(function (resolve) {
+            (function waitForDebuggerClientMessage() {
+                if (
+                    outputConsumer.stateMessages.includes(
```
I plan to chat with you about this tomorrow, it's likely better to have some regex that would ensure it doesn't break us. I don't think there's any risk but there is already some regex in the `_parseDiags` in CMakeOutputConsumer so we might as well.

gcampbell-msft commented on src/diagnostics/cmake.ts:
```diff
@@ -27,6 +27,17 @@ export class CMakeOutputConsumer implements OutputConsumer {
+    get stateMessages() {
+        return this._stateMessages;
+    }
+    private readonly _stateMessages = [] as string[];
```
Changing.

gcampbell-msft commented on src/drivers/cmakeFileApiDriver.ts:
```diff
@@ -245,11 +253,19 @@ export class CMakeFileApiDriver extends CMakeDriver {
             this.configureProcess = child;
+
+            if (withDebugger) {
+                if (outputConsumer instanceof CMakeOutputConsumer) {
```
It shouldn't ever be not a `CMakeOutputConsumer`, but there are methods that call it with a different consumer, though `withDebugger` should always be false in those cases. This also helps please the type checker.

gcampbell-msft commented on package.json:
```diff
@@ -214,6 +214,12 @@
+        "command": "cmake.configureWithDebugger",
+        "title": "%cmake-tools.command.cmake.configureWithDebugger.title%",
+        "category": "CMake",
+        "enablement": "vscode-cmake-tools.cmakeDebuggerAvailable == 'true'"
```
Fixing.

gcampbell-msft commented on src/debug/debuggerConfigureDriver.ts:
```diff
@@ -0,0 +1,26 @@
+        });
+    }
+    await vscode.debug.startDebugging(undefined, {
+        name: "Test CMake Debugger",
```
I don't know if it needs to be localized, but it feels like it can be, I'll localize. And remove Test.

gcampbell-msft replied:
Fixed.

gcampbell-msft commented on src/diagnostics/cmake.ts:
```diff
@@ -73,6 +85,12 @@ export class CMakeOutputConsumer implements OutputConsumer {
     }
 
+    private _parseStateMessages(line: string) {
+        if (line.includes("Waiting for debugger client to connect...")) {
```
Changing.

gcampbell-msft commented on src/debug/debuggerConfigureDriver.ts:
```diff
@@ -0,0 +1,26 @@
+import { CMakeOutputConsumer } from "@cmt/diagnostics/cmake";
+import * as vscode from "vscode";
+
+export const debuggerPipeName = "\\\\.\\\\pipe\\\\cmake-debugger-pipe";
```
Is there a specific format for named pipes in linux/mac?

gcampbell-msft replied:
I just pushed and it may be ready to test in Linux.

gcampbell-msft replied:
Fixed. I haven't explicitly tested on Linux yet.

gcampbell-msft commented:
> One other high-level thought -- since users can now create `launch.json` configurations to debug CMake, I can imagine use cases for debugging in [script mode](https://cmake.org/cmake/help/latest/manual/cmake.1.html#run-a-script) instead of an entire project. Would it be possible to add debug configuration settings to enable this scenario?

Will likely chat about this with you offline

gcampbell-msft commented on src/debug/debuggerConfigureDriver.ts:
```diff
@@ -1,16 +1,23 @@
+    if (process.platform === 'win32') {
+        return `\\\\.\\\\pipe\\\\cmake-debugger-pipe\\\\${uuidv4()}`;
+    } else {
+        return `/cmake/debugger/${uuidv4()}`;
```
Fixing.

gcampbell-msft replied:
Fixed.

gcampbell-msft commented on src/debug/debuggerConfigureDriver.ts:
```diff
@@ -1,16 +1,23 @@
         request: "launch",
-        type: "cmake"
+        type: "cmake",
+        debuggerPipeName
```
Fixing.

benmcmorran commented:
[9682de0](https://github.com/microsoft/vscode-cmake-tools/pull/3181/commits/9682de0cee771147eee389358c1ab180771e5a9f) is still broken on Linux.

benmcmorran commented:
`debuggerDapLog` doesn't seem to have any effect from `launch.json`.

![image](https://github.com/microsoft/vscode-cmake-tools/assets/6981284/5d1637c2-31b1-4203-819c-964501ab6827)

gcampbell-msft commented:
> `debuggerDapLog` doesn't seem to have any effect from `launch.json`.
> 
> ![image](https://user-images.githubusercontent.com/6981284/247986994-5d1637c2-31b1-4203-819c-964501ab6827.png)

Thanks for testing, I've resolved this and added handling of the debugger dap log.

gcampbell-msft commented:
@benmcmorran ping for re-review, I needed to push a small change to fix a build issue.

benmcmorran commented:
You're still out of date with `main`. You'll need to rebase or merge first.

gcampbell-msft commented:
> You're still out of date with `main`. You'll need to rebase or merge first.

I added support for canceling the configuration when the debug session is user cancelled and I rebased with main: 
![VSCodeCancelDebug](https://github.com/microsoft/vscode-cmake-tools/assets/86264750/c0c3a692-bf5e-4926-820a-79e80245db7a)

benmcmorran commented on src/debug/debuggerConfigureDriver.ts:
```diff
@@ -0,0 +1,50 @@
+    }
+}
+
+export class DebugTrackerFactor implements vscode.DebugAdapterTrackerFactory {
```
`DebugTrackerFactory`

gcampbell-msft replied:
Fixed.

benmcmorran commented on src/debug/debuggerConfigureDriver.ts:
```diff
@@ -0,0 +1,50 @@
+                // Despite receiving a message from the debug session, it's not
+                // guaranteed that vscode.debug.activeDebugSession is actually set yet.
+                // Check once per second for up to 10 seconds to see if it shows up.
+                // https://github.com/microsoft/vscode/issues/70125
```
Is all of this needed? It doesn't look like the handler ever accesses the `activeDebugSession`, and by the time the `disconnect` event specifically fires `activeDebugSession` will almost certainly be set anyway. I think the issue this code was designed to fix matters more for startup events.

gcampbell-msft replied:
Good point, I don't think so, fixed.

gcampbell-msft replied:
@benmcmorran sorry for all of the pings, hopefully last one before merge.

Neumann-A commented:
Stupid question: Can this be made a extension separate from configuring `CMakeLists.txt` to debug general cmake scripts invoked with the `-P` option? I want to debug vcpkg portfiles ;)
